### ISSUE TITLE
Configuration of runShadow task appears to be broken by Gradle 6.4.x

### ISSUE THREAD

**Issue by sfitts** [johnrengelman/shadow#574]
### Shadow Version

5.2.0

### Gradle Version

6.4.1

### Expected Behavior

The `runShadow` task requires that `mainClassName` be set (currently after the application of the `application` plug-in and before the application of the `shadow` plug-in).  For example:

```
apply plugin: 'application'
mainClassName = 'io.vantiq.vertx.launcher.VantiqLauncher'
apply plugin: 'com.github.johnrengelman.shadow'
```

Normally when applied in this manner the application runs as expected.  This has been true for all tested versions of Gradle 5.x and is true for Gradle 6.3 (and we presume all prior 6.x releases, though we have not explicitly tested this).

### Actual Behavior

Doing this in Gradle 6.4.1 produces the error:

```
$ ./gradlew servers:standalone:runShadow
> Task :servers:standalone:runShadow FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':servers:standalone:runShadow'.
> The value for task ':servers:standalone:runShadow' property 'mainClass' is final and cannot be changed any further.
```

If I had to guess I'd say this was collateral damage for the changes relating to Java Module support.  In particular you can see that the definition of the JavaApplication type changed from [6.3](https://docs.gradle.org/6.3/dsl/org.gradle.api.plugins.JavaApplication.html) to [6.4.1](https://docs.gradle.org/current/dsl/org.gradle.api.plugins.JavaApplication.html).  Note the removal of `mainClassName` and the introduction of `mainClass`.

This seems like an extremely unfriendly thing to do in a "minor" release and it may well have been unintentional.  But it appears to have completely broken this task.  I can't find a way around this.  Setting `mainClass` in the `application` block as the Gradle documentation suggests leads to the error:

```
* What went wrong:
A problem was found with the configuration of task ':servers:standalone:shadowJar' (type 'ShadowJar').
> No value has been specified for property 'mainClassName'.
```

I have been unable to find something that both sets that value and doesn't violate the fact that it can't be set.

**Comment by ljacomet** [comment]
This is clearly not an intended side effect of the changes that made it in Gradle 6.4 and were tweaked in 6.4.1.

The code in `JavaJarExec` should really be amended to better fit recent Gradle versions. Using `-jar` as the main class name to trick Java in running an executable JAR is clearly bound to cause issues.

**Comment by ljacomet** [comment]
A quick fix for this issue would be to move the `setMain('-jar')` from `JavaJarExec.execute` into `ShadowApplicationPlugin.addRunTask`

**Comment by suushiemaniac** [comment]
Leaving a +1 here as I am running into the exact same issue with Gradle 6.4.1 and Shadow 5.2.0

**Comment by ljacomet** [comment]
> Sadly this really has no workaround and will need a change.

I should be more specific: there is no workaround for making `runShadow` work. It is however possible to define a `JavaExec` task to replace it:

```
tasks.register<JavaExec>("runShadowFixed") {
    val jarFile = (project.tasks.findByName("shadowJar") as Jar).getArchiveFile()
    inputs.files(jarFile).withPropertyName("jarFile").withPathSensitivity(PathSensitivity.RELATIVE)
    setMain("-jar")
    description = "Runs this project as a JVM application using the shadow jar (fixed for Gradle 6.4+)"
    group = ApplicationPlugin.APPLICATION_GROUP
    jvmArgs = application.applicationDefaultJvmArgs.toList()

    doFirst {
        setArgs(mutableListOf(jarFile.get().asFile.path) + getArgs())
    }
}
```

(This snippet uses the Kotlin DSL, conversion to the Groovy one should be simple)

**Comment by bitsnaps** [comment]
> (This snippet uses the Kotlin DSL, conversion to the Groovy one should be simple)

Hi, is this Groovy task equivalent to the Kotlin one?
```groovy
task runShadowFixed(type: JavaExec){
    def jarFile = project.tasks.findByName('shadowJar').asType(Jar).archiveFile
    inputs.files(jarFile).withPropertyName('jarFile').withPathSensitivity(PathSensitivity.RELATIVE)
    setMain('-jar')
    description = 'Runs this project as a JVM app'
    group = ApplicationPlugin.APPLICATION_GROUP
    jvmArgs = application.applicationDefaultJvmArgs.toList()
    doFirst {
        setArgs(jarFile.get().asFile.path.toList() + getArgs())
    }
}
```
